src/io/vss_store.rs

```diff
@@ -0,0 +1,192 @@
+
+use crate::io::utils::check_namespace_key_validity;
+use lightning::util::persist::KVStore;
+use tokio::runtime::Runtime;
+use vss_client::client::VssClient;
+use vss_client::error::VssError;
+use vss_client::types::{
+	DeleteObjectRequest, GetObjectRequest, KeyValue, ListKeyVersionsRequest, PutObjectRequest,
+};
+
+/// A [`KVStore`] implementation that writes to and reads from a [VSS](https://github.com/lightningdevkit/vss-server/blob/main/README.md) backend.
+pub struct VssStore {
+	client: VssClient,
+	store_id: String,
+	runtime: Runtime,
+}
+
+impl VssStore {
+	#[cfg(feature = "vss")]
+	pub(crate) fn new(base_url: &str, store_id: String) -> Self {
+		let client = VssClient::new(base_url);
+		let runtime = tokio::runtime::Builder::new_multi_thread().enable_all().build().unwrap();
+		Self { client, store_id, runtime }
+	}
+
+	fn build_key(&self, namespace: &str, sub_namespace: &str, key: &str) -> io::Result<String> {
+		if key.is_empty() {
+			return Err(Error::new(ErrorKind::Other, "Empty key is not allowed"));
+		}
+		if namespace.is_empty() {
+			Ok(key.to_string())
+		} else {
+			Ok(format!("{}#{}#{}", namespace, sub_namespace, key))
+		}
+	}
+
+	fn split_key(&self, key: &str) -> io::Result<(String, String, String)> {
+		let parts: Vec<&str> = key.split('#').collect();
+		match parts.as_slice() {
+			[namespace, sub_namespace, actual_key] => {
+				Ok((namespace.to_string(), sub_namespace.to_string(), actual_key.to_string()))
+			}
+			_ => Err(Error::new(ErrorKind::InvalidData, "Invalid key format")),
+		}
+	}
+
+	async fn list_all_keys(&self, namespace: &str, sub_namespace: &str) -> io::Result<Vec<String>> {
+		let mut page_token = None;
+		let mut keys = vec![];
+		let key_prefix = format!("{}#{}", namespace, sub_namespace);
+		while page_token != Some("".to_string()) {
+			let request = ListKeyVersionsRequest {
+				store_id: self.store_id.to_string(),
+				key_prefix: Some(key_prefix.to_string()),
+				page_token,
+				page_size: None,
+			};
+
+			let response = self.client.list_key_versions(&request).await.map_err(|e| {
+				let msg = format!("Failed to list keys in {}/{}: {}", namespace, sub_namespace, e);
+				Error::new(ErrorKind::Other, msg)
+			})?;
+
+			for kv in response.key_versions {
+				keys.push(self.split_key(&kv.key)?.2);
+			}
+			page_token = response.next_page_token;
+		}
+		Ok(keys)
+	}
+}
+
+impl KVStore for VssStore {
+	fn read(&self, namespace: &str, sub_namespace: &str, key: &str) -> io::Result<Vec<u8>> {
+		check_namespace_key_validity(namespace, sub_namespace, Some(key), "read")?;
+		let request = GetObjectRequest {
+			store_id: self.store_id.to_string(),
+			key: self.build_key(namespace, sub_namespace, key)?,
+		};
+
+		let resp =
+			tokio::task::block_in_place(|| self.runtime.block_on(self.client.get_object(&request)))
+				.map_err(|e| match e {
+					VssError::NoSuchKeyError(..) => {
+						let msg = format!(
+							"Failed to read as key could not be found: {}/{}/{}. Details: {}",
```

Comment:
Let's word this as `Failed to read from key {}/{}/{}: {}` for consistency with the other errors and since "could not be found" is redundant with both `VssError::NoSuchKeyError` used in the message as `e` and `ErrorKind::NotFound` when wrapped below.

Reply:
Hmm... we don't make such distinctions for other VSS error types. And it _would_ be given first since `ErrorKind::NotFound` comes before the message, no? In practice, I think you'll find that this error message will be extremely verbose and redundant even without the explicit "could not be found" here, IMHO. With our server implementation, the key will even be given again.